### DISCUSSION TITLE
Can PKL be used to validate JSON/YAML documents?

### DISCUSSION THREAD

**Discussion by dtjm** [apple/pkl#195] [Q&A] [answered]
If I have an existing JSON/YAML doc, can it be validated using PKL, or does PKL only support generating JSON/YAML?

**Comment by bioball** [comment, marked as answer]
There's no built-in mechanism for this right now.

There's ways that you _can_ use pkl to do this, where you first convert JSON/YAML to Pkl, then eval the resulting Pkl to produce validation errors. For example, take a look at [kubeval](https://github.com/apple/pkl-pantry/blob/main/packages/pkl.pipe/kubeval.pkl).

We'd like to support this built-in at some point, but it's not on the roadmap at the moment.

**Reply by holzensp** [reply]
You can feed the result of the [JSON](https://pkl-lang.org/package-docs/pkl/current/json/Parser.html) / [YAML](https://pkl-lang.org/package-docs/pkl/current/yaml/Parser.html) parser into [deepToTyped](https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/pkl.experimental.deepToTyped/current/index.html) and render it back out (to make sure everything gets evaluated).

**Reply by kaihendry** [reply]
Would be good to get this on the CLI, as it's a strong use case to validate JSON.
